xds/src/main/java/io/grpc/xds/EnvoyProtoData.java

```diff
@@ -339,4 +339,271 @@ public String toString() {
+    }
+
+    @Override
+    public int hashCode() {
+      return java.util.Objects.hash(destinationPort, prefixRanges, applicationProtocols);
+    }
+
+    @Override
+    public String toString() {
+      return "FilterChainMatch{"
+          + "destinationPort=" + destinationPort
+          + ", prefixRanges=" + prefixRanges
+          + ", applicationProtocols=" + applicationProtocols
+          + '}';
+    }
+  }
+
+  /**
+   * Corresponds to Envoy proto message {@link io.envoyproxy.envoy.api.v2.listener.FilterChain}.
+   */
+  static final class FilterChain {
+    private final FilterChainMatch filterChainMatch;
+    // TODO(sanjaypujare): remove dependency on envoy data type along with rest of the code.
+    private final io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext downstreamTlsContext;
+
+
+    private FilterChain(FilterChainMatch filterChainMatch,
+        io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext downstreamTlsContext) {
+      this.filterChainMatch = filterChainMatch;
+      this.downstreamTlsContext = downstreamTlsContext;
+    }
+
+    static FilterChain fromEnvoyProtoFilterChain(
+        io.envoyproxy.envoy.api.v2.listener.FilterChain proto) {
+      return new FilterChain(
+          FilterChainMatch.fromEnvoyProtoFilterChainMatch(proto.getFilterChainMatch()),
+          proto.getTlsContext()
+      );
+    }
+
+    public FilterChainMatch getFilterChainMatch() {
+      return filterChainMatch;
+    }
+
+    public io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext getDownstreamTlsContext() {
+      return downstreamTlsContext;
+    }
+
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      FilterChain that = (FilterChain) o;
+      return java.util.Objects.equals(filterChainMatch, that.filterChainMatch)
+          && java.util.Objects.equals(downstreamTlsContext, that.downstreamTlsContext);
+    }
+
+    @Override
+    public int hashCode() {
+      return java.util.Objects.hash(filterChainMatch, downstreamTlsContext);
+    }
+
+    @Override
+    public String toString() {
+      return "FilterChain{"
+          + "filterChainMatch=" + filterChainMatch
+          + ", downstreamTlsContext=" + downstreamTlsContext
+          + '}';
+    }
+  }
+
+  /**
+   * Corresponds to Envoy proto message {@link io.envoyproxy.envoy.api.v2.Listener} & related
+   * classes.
+   */
+  static final class Listener {
+    private final String name;
+    private final String address;
+    private final List<FilterChain> filterChains;
+
+    private Listener(String name, String address,
+        List<FilterChain> filterChains) {
+      this.name = name;
+      this.address = address;
+      this.filterChains = filterChains;
+    }
+
+    static String fromEnvoyProtoAddress(io.envoyproxy.envoy.api.v2.core.Address proto) {
```

Comment:
This seems to be a helper method only. Should be private and name it to something different than `fromEnvoyProtoXXX`. Here `fromEnvoyProtoXXX` methods are used as factory methods to create converted data objects. 

Btw, since this helper method is only used once and it's not complex, it can be inlined instead of being a standalone method.

Reply:
Made changes (except inlining because in future we may have an internal Address class)